### ISSUE TITLE
Reporting Not Updating Student count upon removal

### ISSUE THREAD

**Issue by nrherron92** [gocodebox/lifterlms#1596]
### Reproduction Steps

possibly related: https://github.com/gocodebox/lifterlms/issues/1586

+ Add a student to a course
+ check the reporting screen there should be the correct number of students
+ go into that course and remove the student
+ go back into reporting for that course

### Expected Behavior

+ the correct number of students minus one is displayed

### Actual Behavior

+ the student count doesn't account for the removal

### Error Messages / Logs

+ Include any relevant error messages or log files
```
<!-- Paste error logs / backtraces below this line -->

```

### System and Environment Information

<details>
<summary>System Report</summary>

```
Wordpress
-------------------------------------------

Home Url: [removed]
Site Url: [removed]
Login Url: [removed]/wp-login.php
Version: 5.7
Debug Mode: No
Debug Log: No
Debug Display: Yes
Locale: en_US
Multisite: No
Page For Posts: Not Set
Page On Front: Homepage (#155) [[removed]/]
Permalink Structure: /%postname%/
Show On Front: page
Wp Cron: Yes


Settings
-------------------------------------------

Version: 5.0.0-alpha.3
Db Version: 5.0.0-alpha.3
Course Catalog: Course Catalog (#9) [[removed]/courses/]
Membership Catalog: Membership Catalog (#10) [[removed]/memberships/]
Student Dashboard: Dashboard (#12) [[removed]/dashboard/]
Checkout Page: Purchase (#11) [[removed]/purchase/]
Course Catalog Per Page: 9
Course Catalog Sorting: menu_order
Membership Catalog Per Page: 9
Membership Catalog Sorting: menu_order
Site Membership: Not Set
Courses Endpoint: my-courses
Edit Endpoint: edit-account
Lost Password Endpoint: lost-password
Vouchers Endpoint: redeem-voucher
Autogenerate Username: yes
Password Strength Meter: yes
Minimum Password Strength: strong
Terms Required: no
Terms Page: Not Set
Checkout Names: required
Checkout Address: required
Checkout Phone: optional
Checkout Email Confirmation: yes
Open Registration: no
Registration Names: required
Registration Address: optional
Registration Phone: hidden
Registration Voucher: optional
Registration Email Confirmation: no
Account Names: required
Account Address: required
Account Phone: optional
Account Email Confirmation: yes
Confirmation Endpoint: confirm-payment
Force Ssl Checkout: no
Country: US
Currency: USD
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Decimals: 2
Trim Zero Decimals: no
Recurring Payments: yes
Email From Address: [removed]
Email From Name: [removed]
Email Footer Text: 
Email Header Image: 
Cert Bg Width: 800
Cert Bg Height: 616
Cert Legacy Compat: no


Constants
-------------------------------------------

LLMS_REMOVE_ALL_DATA: undefined
LLMS_REST_DISABLE: undefined
LLMS_SITE_FEATURE_RECURRING_PAYMENTS: undefined
LLMS_SITE_IS_CLONE: undefined


Gateways
-------------------------------------------

Authorize.Net: Disabled
Authorize.Net Test Mode: Disabled
Authorize.Net Logging: 
Authorize.Net Order: 1
PayPal: Disabled
PayPal Test Mode: Disabled
PayPal Logging: 
PayPal Order: 1
Stripe: Enabled
Stripe Test Mode: Enabled
Stripe Logging: no
Stripe Order: 1
Manual: Disabled
Manual Logging: 
Manual Order: 1


Server
-------------------------------------------

Mysql Version: 5.7.33
Php Curl: Yes
Php Default Timezone: UTC
Php Fsockopen: Yes
Php Max Input Vars: 5000
Php Max Upload Size: 512 MB
Php Memory Limit: 256M
Php Post Max Size: 1024M
Php Soap: Yes
Php Suhosin: No
Php Time Limt: 30
Php Version: 7.3.27
Software: Apache/2.4.46 (Unix) OpenSSL/1.1.1
Wp Memory Limit: 256M


Browser
-------------------------------------------

HTTP USER AGENT: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.128 Safari/537.36/noSbTd0P-2


Theme
-------------------------------------------

Name: Genesis Sample
Version: 3.4.1
Themeuri: https://demo.studiopress.com/
Authoruri: https://www.studiopress.com/
Template: genesis
Child Theme: Yes
Llms Support: No


Plugins
-------------------------------------------

LifterLMS: 5.0.0-alpha.3
LifterLMS Advanced Quizzes: 1.1.1
LifterLMS Advanced Videos: 1.0.0-beta.14
LifterLMS Assignments: 1.1.12
LifterLMS Authorize.Net Payment Gateway: 1.1.0
LifterLMS Custom Fields: 1.0.2
LifterLMS Groups: 1.0.0-beta.10
LifterLMS PayPal Gateway: 1.3.0
LifterLMS PDFs: 1.0.0
LifterLMS Private Areas: 1.1.2
LifterLMS Social Learning: 1.3.0
LifterLMS Stripe Payment Gateway: 5.3.2
LifterLMS WooCommerce: 2.1.3
User Switching: 1.5.6


Integrations
-------------------------------------------

BbPress: No
BuddyPress: No
LifterLMS Groups: Yes
WooCommerce: No
LifterLMS Private Areas: Yes
LifterLMS Social Learning: Yes
Videos: Vimeo: No
Videos: Wistia: No
Videos: YouTube: No


Template Overrides
-------------------------------------------




```

</details>

This issue has be recreated:
+ [X] Locally
+ [X] On a staging site
+ [X] On a production website
+ [X] With only LifterLMS and a default theme

### Browser, Device, and Operating System Information

+ Browser name and version
+ Operating System name and version
+ Device name and version (if applicable)

HS-163174

**Comment by thomasplevy** [comment]
@nrherron92 this is not at all related to #1586, instead this is related to #1564

Since we're now caching student counts as a performance improvement it's likely that on larger sites reporting data may not be 100% accurate all the time.

Please confirm that background processes are running as expected (which it's possible they're not running as quickly as users would like if they have staging sites) and let me know if you need more help here.

In short: on small sites student counts should be accurate next to real time but not 100% accurate 100% of the time and if the site is large (there's 500+ students in a course) the number will be accurrate within 4 hours when the background process runs.

As @eri-trabiccolo discussed in #1564 this is something of a bug/regression brought about by performance improvements to reduce queries (I added logic to cache the student count data in order to cut down on the number of expensive queries run to determine the number of students in a course). I created an issue here and we're now aware of it but as of yet have not prioritized a fix.

This will be fixed in a future release but as of this moment we're probably not going to make any changes here until after version 5.x is moved to public beta (so hopefully we'll get on this within the next two weeks)

**Comment by nrherron92** [comment]
@thomasplevy is it possible that we don’t update the enrollments number when the course has no students? I did just go back and check this on a trial site and you're right the numbers update in real time for a low number of students (which the user also has) but if you say enroll 5 students and then go through and remove all five of those students from the course so it has 0 students, the number never goes down to zero. 

if you remove them one at a time the lowest it goes to is 1

**Comment by eri-trabiccolo** [comment]
Yeah as @nrherron92 says we sure have a problem when a course gets to have no students, because we update the cached enrolled students number in the `task_complete()` (as well as the other cached data) 
https://github.com/gocodebox/lifterlms/blob/6c2c4a9bb578a67d65682122d325d984c4be0dbe/includes/processors/class.llms.processor.course.data.php#L464

But we don't execute that method if the course has no students
https://github.com/gocodebox/lifterlms/blob/6c2c4a9bb578a67d65682122d325d984c4be0dbe/includes/processors/class.llms.processor.course.data.php#L98-L100

**Comment by thomasplevy** [comment]
@nrherron92 is the problem the user is having when there's only a single student in the course?

I've written a patch to fix this particular scenario (#1603) but even if this patch would satisfy the user I think we should leave this open to further improve alongside #1564
